codex-rs/tui/src/chatwidget.rs

```diff
@@ -999,6 +1005,15 @@ impl ChatWidget {
         let UserMessage { text, image_paths } = user_message;
         let mut items: Vec<InputItem> = Vec::new();
 
+        // Special-case: "!cmd" executes a local shell command instead of sending to the model.
+        if let Some(stripped) = text.strip_prefix('!') {
+            let cmd = stripped.trim().to_string();
+            if !cmd.is_empty() {
```

Comment:
So if the user types `!` by itself, we'll send the message `"!"` to the agent, correct?

I guess this is fine, though we could take this opportunity to tell the user about the `!` feature instead?